crates/matrix-sdk/src/authentication/qrcode/rendezvous_channel.rs

```diff
@@ -0,0 +1,433 @@
+    /// device.
+    ///
+    /// The message must be of the `text/plain` content type.
+    #[instrument(skip_all)]
+    pub(super) async fn send(&mut self, message: Vec<u8>) -> Result<(), HttpError> {
+        let etag = self.etag.clone();
+
+        let request = self
+            .client
+            .inner
+            .request(Method::PUT, self.rendezvous_url().to_owned())
+            .body(message)
+            .header(IF_MATCH, etag)
+            .header(CONTENT_TYPE, TEXT_PLAIN_CONTENT_TYPE);
+
+        debug!("Sending a request to the rendezvous channel {request:?}");
+
+        let response = request.send().await?;
+        let status = response.status();
+
+        debug!("Response for the rendezvous sending request {response:?}");
+
+        if status.is_success() {
+            // We successfully send out a message, get the ETAG and update our internal copy
+            // of the ETAG.
+            let etag = get_header(response.headers(), &ETAG)?;
+            self.etag = etag;
+
+            Ok(())
+        } else {
+            let builder = http::Response::builder().status(status);
+            let body = response.bytes().await?;
+
+            let response = builder.body(body).map_err(IntoHttpError::from)?;
+            let error = FromHttpResponseError::<RumaApiError>::Server(RumaApiError::Other(
+                MatrixError::from_http_response(response),
+            ))
+            .into();
+
+            Err(error)
+        }
+    }
+
+    /// Attempt to receive a message from the [`RendezvousChannel`] from the
+    /// other device.
+    ///
+    /// The should be of the `text/plain` content type but the parsing and
+    /// verification of this fact is left up to the caller.
+    ///
+    /// This method will wait in a loop for the channel to give us a new
+    /// message.
+    pub(super) async fn receive(&mut self) -> Result<Vec<u8>, HttpError> {
+        loop {
+            match self.receive_single_message().await {
+                Ok(message) => {
+                    trace!(
+                        status_code = %message.status_code,
+                        "Received data from the rendezvous channel"
+                    );
+
+                    if message.status_code == StatusCode::OK
+                        && &message.content_type == TEXT_PLAIN_CONTENT_TYPE
+                        && message.body.len() != 0
+                    {
+                        return Ok(message.body);
+                    }
+                }
+                Err(e) => {
+                    warn!("Error while receiving data from the rendezvous channel {e:?}");
+
+                    // If it's a permanent error, return an error.
+                    if let Some(kind) = e.client_api_error_kind() {
+                        if *kind == ErrorKind::NotFound {
+                            return Err(e.into());
+                        }
+                    }
```

Comment:
Can we add a maximum number of “retries” here so that we are sure we will leave this `loop` whatever happens? For example, if the server is busy, it will return a 5xx error, which is not a 404 as expected here, in which case we will never escape this loop.

Reply:
The server will expire the rendezvous session, and if it doesn't do so we should probably expect the user to cancel things.

But you are right, this is a bit too optimistic, we should forward other error types to the user as well.

Fixed and added some more test for this in: 74883c816e129fa2f643bf74c9e470a0a13491fb.